### ISSUE TITLE
DKMS compilations fails on 6.3

### ISSUE THREAD

**Issue by emansom** [cyring/CoreFreq#422]
When using [linux-cachyos-rc](https://aur.archlinux.org/packages/linux-cachyos-rc) as kernel, DKMS compilation fails.

```
/var/lib/dkms/corefreq/1.95.5/build/corefreqk.c: In functie ‘CoreFreqK_mmap’:
/var/lib/dkms/corefreq/1.95.5/build/corefreqk.c:22084:23: fout: assignment of read-only member ‘vm_flags’
22084 |         vma->vm_flags = VM_READ;
      |                       ^
compilatie is beëindigd ten gevolge van -Wfatal-errors.
make[2]: *** [scripts/Makefile.build:252: /var/lib/dkms/corefreq/1.95.5/build/corefreqk.o] Fout 1
make[1]: *** [Makefile:2026: /var/lib/dkms/corefreq/1.95.5/build] Fout 2
make[1]: Map '/usr/lib/modules/6.3.0-rc1-1-cachyos-rc/build' wordt verlaten
make: *** [Makefile:86: all] Fout 2
make: Map '/usr/src/corefreq-1.95.5' wordt verlaten
```

**Comment by emansom** [comment]
Builds and works perfectly now!

**Comment by emansom** [comment]
When build with `ARCH_PMC=UMC` the `corefreq-cli` TUI errors out on launch:

```
Daemon connection error code 2
corefreq-ro-shm: 'Bestand of map bestaat niet' @ line 21307
```

Is there some setup required for UMC?

**Comment by emansom** [comment]
Daemon is not launching.

```
$ sudo corefreqd -d
Driver connection error code 13
Version 0.0.0: 'Permission denied' @ line 9213
```

**Comment by emansom** [comment]
> And I need your acknowledgement if there's no remaining issue before merging into master branch. Thank you.

The computer with the 3500X is my parents' HTPC. I don't have access to it all the time (_remotely, reverse tunnel via site-to-site VPN_) as it's not always powered on.

I'll be at their place today for some maintenance on the HTPC and try to debug further.

PBO and UEFI settings are heavily for low power usage and low thermal envelope, including disablement of cores (2 of the 6 are enabled). I suspect the failures may have something to do with that.

I'll check what dmesg prints after loading the module. Is there any extra compilation flags or module options to increase its verbosity?

**Comment by emansom** [comment]
> `dmesg` will be enough to diagnose `corefreqk.ko` if any issue.
> Have also a look in modules list with `lsmod` to check if the version aligned driver is well loaded.

Only thing it prints is some info, no error.

```console
ewout@enthoo ~ % sudo dmesg | tail
[  230.475320] rfkill: input handler disabled
[  231.205749] input: solaar-keyboard as /devices/virtual/input/input31
[ 1162.036175] logitech-djreceiver 0003:046D:C52B.0004: device of type eQUAD step 4 DJ (0x04) connected on slot 2
[ 1162.058254] input: Logitech K270 as /devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-1/3-1:1.2/0003:046D:C52B.0004/0003:046D:4003.0009/input/input32
[ 1162.062013] logitech-hidpp-device 0003:046D:4003.0009: input,hidraw3: USB HID v1.11 Keyboard [Logitech K270] on usb-0000:07:00.3-1/input2:2
[ 1162.176015] logitech-hidpp-device 0003:046D:4003.0009: HID++ 2.0 device connected.
[ 3171.121897] snd_hda_codec_hdmi hdaudioC0D0: HDMI: audio coding xtype 11 not expected
[ 5079.763522] overlayfs: "xino" feature enabled using 2 upper inode bits.
[ 5079.831385] overlayfs: "xino" feature enabled using 2 upper inode bits.
[ 6396.239435] CoreFreq(1:-1:-1): Processor [ 8F_71] Architecture [Zen2/Matisse] CPU [2/2]
ewout@enthoo ~ % lsmod | grep -i corefreqk
corefreqk             622592  0
ewout@enthoo ~ %
```

> Starting Daemon in debug mode will help you check if one monitoring loop thread is running per CPU. Thus, start with `corefreqd -d`

```
ewout@enthoo ~ % sudo corefreqd -d
Driver connection error code 13
Version 0.0.0: 'Permission denied' @ line 9213
```

🤷‍♂️

**Comment by emansom** [comment]
> SMT also appears disabled. Not recommended because some features are deactivated like S3 SuspendToRam, other registers...

The [3500X](https://www.amd.com/en/products/cpu/amd-ryzen-5-3500x) is a 3600 with disabled SMT from factory, can't enable it.

> To minimize the consumed power I won't touch the Core count and left enabled those BIOS settings. Same with SMT. I will make sure all C-States, P-States, DVFS are enabled, preferably AUTO.
> 
> PL, **Power Limiter**, is what you should tune. ASUS provides an easy ECO mode: down to 65W on my 3950X. But you loose horse power if that don't mind.

The PC is used primarily as HTPC, just watching online content. The build uses a NH-L9x65 and its stock configuration the cooler couldn't keep up. With a custom "35W Eco Mode" (PPT = 40W, TDC = 35A, EDC = 40A) all is fine again. The disablement of cores was done to ensure it never ramps up any fans.

> You can also tune the Vcore, SoC, and DRAM voltage

XMP is disabled, that's about it on underclocking.

> Think also that kernel will migrate at its best all tasks to the two total CPUs. It's bad balance to my taste. I would rather enabled all Cores, including SMT, and take into account the performance score of each CPU. See the CPPC ratios listed in _CoreFreq_ With that knowledge you can then map efficiently the task-cpu affinity. Containers, Virtualization can also make use of power affinity.

In my testing C6 state wasn't toggled much in those scenarios, when playing DRM encrypted videos in Firefox (no hardware acceleration support in Widevine on Linux, so CPU decryption + H264 decode). Having only two cores enabled is more energy efficient.

**Comment by emansom** [comment]
> So far, the Daemon issue is due to unsupported factors.

Yes it seems not related to 6.3 indeed. Probably should migrate to a new issue. Thanks for trying to reproduce 👍🏻

**Comment by gel-crabs** [comment]
[journalctl.txt](https://github.com/cyring/CoreFreq/files/10986103/journalctl.txt)

Here's what occurs when trying to run CoreFreq on 6.3 with the latest commits

**Comment by gel-crabs** [comment]
[journalctl.txt](https://github.com/cyring/CoreFreq/files/10993832/journalctl.txt)

corefreqd -d freezes and hangs the console. This is with or without ARCH_PMC=UMC.

**Comment by gel-crabs** [comment]
Yes, I commented them out as requested above and rebuilt it.

**Comment by gel-crabs** [comment]
linux-cachyos-rc 6.3-rc2

[config.txt](https://github.com/cyring/CoreFreq/files/10995128/config.txt)

**Comment by gel-crabs** [comment]
I'm currently compiling it without the per-VMA locks patch to see if that's the problem

**Comment by gel-crabs** [comment]
It is not the per-VMA locks, I'm going to compile it without any patches now

**Comment by gel-crabs** [comment]
I also noticed this line in my dmesg:

[    2.298642] memfd_create() without MFD_EXEC nor MFD_NOEXEC_SEAL, pid=1 'systemd'

**Comment by gel-crabs** [comment]
I think that's unrelated. I still get the same error without any patches; it seems to have something to do with vm_mmap_pgoff. I'm going to try it with GCC and O2 now

**Comment by gel-crabs** [comment]
Yep, it was Clang. It's working now.

**Comment by gel-crabs** [comment]
> > Yep, it was Clang. It's working now.
> > Clang, -flto=auto, and -O3
> 
> Thank you. Somehow it has to be documented into Readme.
> 
> @emansom : Were you also building kernel with clang while Daemon being denied ?

It was working up to this point, it's most likely to do with this being an RC kernel; these kinds of issues usually get fixed by the time the kernel gets released.

**Comment by emansom** [comment]
> @emansom : Were you also building kernel with clang while Daemon being denied ?

Not sure which compiler is triggered by the [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=corefreq) (_modified locally to pull from the linux_6_3 branch_).

I'd guess it defaulted to gcc, not clang/llvm.

**Comment by ppascher** [comment]
Sorry to bring this up again but I just build linux-cachyos 6.3.0-3 from AUR and ran into the same issue:
```
cat /proc/version 
Linux version 6.3.0-3-cachyos (linux-cachyos@cachyos) (gcc (GCC) 12.2.1 20230201, GNU ld (GNU Binutils) 2.40) #1 SMP PREEMPT_DYNAMIC Wed, 26 Apr 2023 15:47:05 +0000
```

Compiling corefreq DKMS errors with:
```
/var/lib/dkms/corefreq/1.95.5/build/corefreqk.c: In function ‘CoreFreqK_mmap’:
/var/lib/dkms/corefreq/1.95.5/build/corefreqk.c:22084:23: error: assignment of read-only member ‘vm_flags’
22084 |         vma->vm_flags = VM_READ;
      |                       ^
compilation terminated due to -Wfatal-errors.
```
And kernel config shows:
```
zcat /proc/config.gz | grep CONFIG_FORT
CONFIG_FORTIFY_SOURCE=y
```
The [latest commit](https://github.com/CachyOS/linux-cachyos/commit/be42265b811d2bf362d1ec87d9ba72dcf5b9352a) to the cachyos AUR package reads `Disable per-VMA lock`
Could this be the reason?
As I understand I should not run into this issue as my kernel is built with gcc?
Thank you.
